### PR TITLE
Os-199 Schema configurable audit fields

### PR DESCRIPTION
**Pull request by pritha-tarento**
Introduced AuditFields which does auditing for each new added entity and updated entity.
AuditFields: enum values are oscreatedby, oscreatedat, osupdatedby, osupdatedat.

Schema holds the list of field names which are expected to be stored. 
The audit is schema driven which can be enabled or disable as per need.

### CODE REVIEW

File: java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuditHelper {
+
+    private static Logger logger = LoggerFactory.getLogger(RegistryServiceImpl.class);
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java]: Check class name.

File: java/middleware-commons/src/main/java/io/opensaber/registry/middleware/util/JSONUtil.java [hunk 2]
```diff
@@ -199,6 +199,34 @@ public static void addNode(ObjectNode parent, String childKey, ObjectNode child)
...
+            if (entryValue.isArray()) {
+                for (int i = 0; i < entryValue.size(); i++) {
+                    if (entry.getValue().get(i).isObject())
+                        addField((ObjectNode) entry.getValue().get(i), firstFieldName, childKey, child);
+                }
+            }
+        });
+        if (childKey == null || childKey.isEmpty())
+            throw new IllegalArgumentException(KEY_NULL_ERROR);
+        if(!parent.fieldNames().next().equalsIgnoreCase(firstFieldName))
```
Comment (indrajra) [inline java/middleware-commons/src/main/java/io/opensaber/registry/middleware/util/JSONUtil.java]: need for firstFieldName? This way of checking is error-prone. We must not do this.

File: java/registry/src/main/resources/public/_schemas/Teacher.json [hunk 3]
```diff
@@ -149,6 +149,7 @@
          "privateFields": ["nationalIdentifier", "teacherCode", "birthDate"],
          "signedFields": ["serialNum"],
          "indexFields": ["(serialNum, teacherCode)","teacherName"],
-         "uniqueIndexFields": ["serialNum", "teacherCode"]
+         "uniqueIndexFields": ["serialNum", "teacherCode"],
+         "systemFields": ["_osCreatedAt", "_osUpdatedAt", "_osCreatedBy"]
```
Comment (indrajra) [inline java/registry/src/main/resources/public/_schemas/Teacher.json]: Comments are not updated.
Please put this for other examples - Vehicle and Person also.

File: java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java (hunk 1 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java]: Ok, done

File: java/middleware-commons/src/main/java/io/opensaber/registry/middleware/util/JSONUtil.java (hunk 2 above)
Comment (pritha-tarento) [reply java/middleware-commons/src/main/java/io/opensaber/registry/middleware/util/JSONUtil.java]: removed firstFieldName. Done

File: java/registry/src/main/resources/public/_schemas/Teacher.json (hunk 3 above)
Comment (pritha-tarento) [reply java/registry/src/main/resources/public/_schemas/Teacher.json]: Done

File: java/registry/src/test/java/io/opensaber/registry/util/OSSystemFieldsHelperTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    }
+
+    @Test
+    public void testAddSystemPropertyUpdatedBy() throws IOException {
+        JsonNode testNode = getTestNode();
+        String key = testNode.fieldNames().next();
+        JsonNode node = testNode.get(key);
+
+        systemFieldsHelper.addSystemProperty("_osUpdatedBy", node, "userId", "timeStamp");
```
Comment (indrajra) [inline java/registry/src/test/java/io/opensaber/registry/util/OSSystemFieldsHelperTest.java]: what's the use of timeStamp here? Maybe we could pass null.

Comment (pritha-tarento) [reply java/registry/src/test/java/io/opensaber/registry/util/OSSystemFieldsHelperTest.java]: Null as empty could have been passed value. I choose to pass a string (named:timestamp) So i know in test data what is this for.

**Pull request merged**